components/brave_wallet/renderer/js_ethereum_provider.cc

```diff
@@ -177,19 +197,44 @@ void JSEthereumProvider::Install(bool allow_overwrite_window_ethereum_provider,
     return;
   }
   v8::Local<v8::Value> provider_value = provider.ToV8();
+  v8::Local<v8::Object> provider_object =
+      provider_value->ToObject(context).ToLocalChecked();
+
+  // Create a proxy to the actual JSEthereumProvider object which will be
+  // exposed via window.ethereum.
+  // This proxy uses a handler which would call things directly on the actual
+  // JSEthereumProvider object so dApps which creates and uses their own proxy
+  // of window.ethereum to access our provider won't throw a "Illegal
+  // invocation: Function must be called on an object of type
+  // JSEthereumProvider" error.
+  blink::WebLocalFrame* web_frame = render_frame->GetWebFrame();
+  v8::Local<v8::Proxy> ethereum_proxy;
+  auto ethereum_proxy_handler_val = ExecuteScript(
+      web_frame, kEthereumProxyHandlerScript, kEthereumProxyScript);
+  v8::Local<v8::Object> ethereum_proxy_handler_obj =
+      ethereum_proxy_handler_val.ToLocalChecked()
+          ->ToObject(context)
+          .ToLocalChecked();
+  if (!v8::Proxy::New(context, provider_object, ethereum_proxy_handler_obj)
+           .ToLocal(&ethereum_proxy)) {
+    return;
+  }
+
+  // The actual JSEthereumProvider object is at window.braveEthereum which we
+  // would run defineProperties on in ethereum_provider.js to make something
+  // readonly, removeAllListeners for example.
+  SetProviderNonWritable(context, global, provider_value,
+                         gin::StringToV8(isolate, kBraveEthereum), true);
```

Comment:
we can actually use `window.ethereum` directly without export an extra provider. In this way, iOS doesn't need extra change to take care of `window.braveEthereum`